pkg/window/strategy/sliding/sliding.go

```diff
@@ -0,0 +1,168 @@
+	// list.List is implemented as a doubly linked list which allows us to traverse the nodes in
+	// both the directions.
+	// Although the worst case time complexity is O(n), because of the time based ordering and
+	// since the elements are rarely out of order, the amortized complexity works out to be closer to O(1)
+	// Because most of the keys are expected to be associated with the most recent window, we always start
+	// the traversal from the tail of the list for Get and Create Operations. For Remove Operations, since
+	// the earlier windows are expected to be closed before the more recent ones, we start the traversal
+	// from the Head.
+	entries *list.List
+	lock    sync.RWMutex
+}
+
+var _ window.Windower = (*Sliding)(nil)
+
+// NewSliding returns a Sliding windower
+func NewSliding(length time.Duration, slide time.Duration) *Sliding {
+	return &Sliding{
+		Length:  length,
+		Slide:   slide,
+		entries: list.New(),
+		lock:    sync.RWMutex{},
+	}
+}
+
+// AssignWindow returns a set of windows that contain the element based on event time
+func (s *Sliding) AssignWindow(eventTime time.Time) []window.AlignedKeyedWindower {
+	// start time of the window in to which this element certainly belongs.
+	startTime := eventTime.Truncate(s.Length)
+	// end time of the window in to which this element certainly belongs.
+	endTime := startTime.Add(s.Length)
+
+	// we have to find the boundaries of the sliding windows that are possible
+	// first lets consider end time as fixed, find the min start and end times
+	minEndTime := startTime.Add(s.Length % s.Slide)
+	minStartTime := minEndTime.Add(-s.Length)
+
+	// lets consider start time as fixed and find the max start and end times.
+	maxStartTime := endTime.Add(-(s.Length % s.Slide))
+	maxEndTime := maxStartTime.Add(s.Length)
+
+	// now all the windows should fall in between maxend and minend times.
+	// one could consider min start and max start times as well.
+	wCount := int((maxEndTime.Sub(minEndTime)) / s.Slide)
+	windows := make([]window.AlignedKeyedWindower, 0)
+
+	for i := 0; i < wCount; i++ {
+		// we make the windows left aligned since the original truncation operation
+		// is left aligned.
+		st := minStartTime.Add(time.Duration(i) * s.Slide)
+		et := st.Add(s.Length)
+
+		// since there is overlap at the boundaries
+		// we attribute the element to the window to the right (higher)
+		// of the boundary
+		// left exclusive and right inclusive
+		// so given windows 500-600 and 600-700 and the event time is 600
+		// we will add the element to 600-700 window and not to the 500-600 window.
+		if eventTime.Before(st) || !eventTime.Before(et) {
+			continue
+		}
+
+		akw := keyed.NewKeyedWindow(st, et)
+
+		windows = append(windows, akw)
+	}
+
+	return windows
+}
+
+// InsertIfNotPresent inserts a window to the list of active windows if not present and returns the window
+func (s *Sliding) InsertIfNotPresent(kw window.AlignedKeyedWindower) (aw window.AlignedKeyedWindower, isPresent bool) {
```

Comment:
since we need to invoke this function for every message, we might have to think about caching this. we can make that decision after a performance analysis.

Reply:
Okayy